feat: add PostHog analytics integration

amitsharm opened the pull request:
- Add PostHog client configuration in ui/lib/posthog.ts
- Integrate PostHog tracking in auth forms and provider workflows
- Add PostHog dependencies (posthog-js and posthog-node)
- Update Next.js config for PostHog integration
- Add environment variables for PostHog configuration
- Update changelog with PostHog integration entry

### Context

Please include relevant motivation and context for this PR.

If fixes an issue please add it with `Fix #XXXX`

### Description

Please include a summary of the change and which issue is fixed. List any dependencies that are required for this change.

### Checklist

- Are there new checks included in this PR? Yes / No
    - If so, do we need to update permissions for the provider? Please review this carefully.
- [ ] Review if the code is being covered by tests.
- [ ] Review if code is being documented following this specification https://github.com/google/styleguide/blob/gh-pages/pyguide.md#38-comments-and-docstrings
- [ ] Review if backport is needed.
- [ ] Review if is needed to change the [Readme.md](https://github.com/prowler-cloud/prowler/blob/master/README.md)
- [ ] Ensure new entries are added to [CHANGELOG.md](https://github.com/prowler-cloud/prowler/blob/master/prowler/CHANGELOG.md), if applicable.

#### API
- [ ] Verify if API specs need to be regenerated.
- [ ] Check if version updates are required (e.g., specs, Poetry, etc.).
- [ ] Ensure new entries are added to [CHANGELOG.md](https://github.com/prowler-cloud/prowler/blob/master/api/CHANGELOG.md), if applicable.

### License

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

Commit:
feat: add PostHog analytics integration

- Add PostHog client configuration in ui/lib/posthog.ts
- Integrate PostHog tracking in auth forms and provider workflows
- Add PostHog dependencies (posthog-js and posthog-node)
- Update Next.js config for PostHog integration
- Add environment variables for PostHog configuration
- Update changelog with PostHog integration entry

cesararroba commented on .env:
```diff
@@ -150,3 +150,7 @@ LANGSMITH_TRACING=false
+
+# posthog integration
+NEXT_PUBLIC_POSTHOG_KEY=""
+NEXT_PUBLIC_POSTHOG_HOST=https://eu.i.posthog.com
```
```suggestion
NEXT_PUBLIC_POSTHOG_HOST=""
```
I would leave this blank as well, user can use a different endpoint

amitsharm replied:
Done. changed the host to blank

Commit:
fix: update package-lock.json to sync with PostHog dependencies

Commit:
fix: let users bring their own posthog host

paabloLC commented on ui/app/providers.tsx:
```diff
@@ -15,10 +18,24 @@ export interface ProvidersProps {
 export function Providers({ children, themeProps }: ProvidersProps) {
   const router = useRouter();
 
+  useEffect(() => {
```
let's avoid to use `useEffect` here, please

paabloLC commented on ui/components/auth/oss/auth-form.tsx:
```diff
@@ -128,6 +136,20 @@ export const AuthForm = ({
+          lastName = nameParts.slice(1).join(" ") || "";
+        }
+        posthog.capture("user_registered", {
+          email: data.email.toLocaleLowerCase(),
```
for what is `toLocaleLowerCase` needed?

amitsharm replied:
converting to lower case in analytics.ts
 email: email.toLowerCase(),

paabloLC commented on ui/components/auth/oss/auth-form.tsx:
```diff
@@ -128,6 +136,20 @@ export const AuthForm = ({
+          firstName = nameParts[0] || "";
+          lastName = nameParts.slice(1).join(" ") || "";
+        }
+        posthog.capture("user_registered", {
```
we're using camelCase to named vars

amitsharm replied:
done.

paabloLC commented on ui/components/auth/oss/auth-form.tsx:
```diff
@@ -107,6 +110,11 @@ export const AuthForm = ({
       });
       if (result?.message === "Success") {
+        posthog.identify(data.email.toLowerCase());
+        posthog.capture("user_login", {
```
we're using camelCase to named vars

amitsharm replied:
changed to userLogin

paabloLC commented on ui/lib/posthog.ts:
```diff
@@ -0,0 +1,11 @@
+import { PostHog } from "posthog-node";
+
+// NOTE: This is a Node.js client, so you can use it for sending events from the server side to PostHog.
+export default function PostHogClient() {
```
we're using arrow functions

paabloLC commented on ui/components/providers/workflow/forms/test-connection-form.tsx:
```diff
@@ -145,6 +146,13 @@ export const TestConnectionForm = ({
             } else {
               setIsRedirecting(true);
+              // Capture PostHog event for successful cloud connection
+              posthog.capture("cloud_connection_success", {
```
do we know if this is working?

amitsharm replied:
it is working.  also changed to camel case

paabloLC reviewed (changes requested):
I've been working on the PostHog integration and realized we should clean this up a bit. Right now all the analytics logic is mixed into the auth form component, which makes it harder to maintain and reuse.

Could you help me refactor this? The idea is to create a new file `lib/analytics.ts` that centralizes all our PostHog tracking functions. We'd have functions like `trackUserLogin()` and `trackUserRegistration()` that handle all the PostHog calls internally with proper error handling.

Then in the auth form component, instead of having all that PostHog code inline, we'd just call these clean functions. For example, when a user logs in successfully, instead of calling `posthog.identify()` and `posthog.capture()` directly, we'd just call `trackUserLogin({ email: data.email })`.

The main benefits would be that the auth component stays focused on UI logic, we can reuse these tracking functions elsewhere, and if we need to change how we track events, we only have to update one place.

I've already worked out the structure, we'd need TypeScript interfaces for the event data, proper error handling so tracking failures don't break the user flow, and a utility function to parse names into first/last name parts.

I also leave some comments in line directly in the code.

Thanks

alejandrobailo reviewed (changes requested):
There are a few missing points about the implementation:

[1. Server-side configuration](https://posthog.com/docs/libraries/next-js#server-side-configuration)

[2. Configuring a reverse proxy to PostHog](https://posthog.com/docs/libraries/next-js#configuring-a-reverse-proxy-to-posthog)

<img width="550" height="470" alt="Image" src="https://github.com/user-attachments/assets/c67bf98c-3233-414c-b84f-0d4339af852b" />

Commit:
feat: centralize PostHog analytics with proper error handling

- Create lib/analytics.ts with centralized tracking functions
- Use arrow functions and camelCase naming per code standards
- Add error handling to prevent analytics failures from crashing app
- Update auth forms and provider workflows to use new analytics lib
- Export TypeScript interfaces for type-safe analytics payloads

Addresses PR review comments:
- Convert to arrow functions (requested by @paabloLC)
- Use camelCase for event and property names
- Centralize analytics logic for better maintainability

amitsharm committed:
Update ui/CHANGELOG.md

Co-authored-by: Alejandro Bailo <59607668+alejandrobailo@users.noreply.github.com>

Commit:
fix: resolved linting and prettier issues for formating

Commit:
fix: updated per @paabloLC request to move Posthog initialization outside of providers.tsx

Commit:
fix: resolved linting errors

alejandrobailo commented on ui/app/providers.tsx:
```diff
@@ -5,6 +5,9 @@ import { useRouter } from "next/navigation";
 import { ThemeProvider as NextThemesProvider } from "next-themes";
 import { ThemeProviderProps } from "next-themes/dist/types";
+import posthog from "posthog-js";
+import { PostHogProvider as PHProvider } from "posthog-js/react";
```
Okay, I suppose you've tested this if this implementation works. I'm afraid of this "This implementation is for Next.js versions above 15; we are on the 14.X"

jfagoagas commented on ui/lib/analytics.ts:
```diff
@@ -0,0 +1,173 @@
+  if (typeof window === "undefined") return; // Don't initialize on server side
+
+  try {
+    posthog.init(process.env.NEXT_PUBLIC_POSTHOG_KEY!, {
```
@amitsharm I've reviewed this from our last conversation https://posthog.com/docs/session-replay/privacy

We need to start by ```
{
    maskAllInputs: true,
    maskTextSelector: "*"
}
``` and then unmask some other inputs if we need to, to comply with our data retention policy.

I recommend also to pass this configuration from the environment to have a quick way to update it without changing code.

If not we'll need to setup something to work with our tenants lifecycle as data must be deleted from all our systems and 3rd parties within 10 calendar days https://docs.prowler.com/security/#data-retention-policy

amitsharm replied:
Added, we do not nee to see customer inputs so  maskAllInputs: true, is added. These is the term of conditions for Prowler Cloud. 
https://prowler.com/terms-of-service/

jfagoagas replied:
Great! Did you forget to push the changes?

amitsharm replied:
ooh.. just pushed the change and requested review.

jfagoagas commented on ui/components/providers/workflow/forms/test-connection-form.tsx:
```diff
@@ -145,6 +146,12 @@ export const TestConnectionForm = ({
             } else {
               setIsRedirecting(true);
+              // Track cloud connection success event
+              trackCloudConnectionSuccess({
```
Why not to create a PostHog wrapper to centralize the calls like:

```
track_event(event_type: "cloud_connection_success", payload: "{}")
```

Because today we use PostHog and who knows what other things we'll use.

That way we'll have an easy way to identify all places where we are tracking events. Furthermore, all evens can be created as an enum.

amitsharm replied:
added a generic event tracking function trackEvent in analytics.ts that allows capturing custom events in with any name and properties. We can use this function when we need to capture events.

Commit:
fix: updated posthog init function to mask everything in session replays

Commit:
fix: updated posthog init function to mask everything in session replays

jfagoagas commented on .env:
```diff
@@ -152,6 +152,5 @@ LANGSMITH_API_KEY=""
-NEXT_PUBLIC_POSTHOG_HOST=""
-NEXT_PUBLIC_POSTHOG_KEY=""
-
+NEXT_PUBLIC_POSTHOG_KEY=phc_dHS2LOFJ3gJWVt8kP6ruQgSTDuSAfRDStv6eMUMeMBt
```
Watch out!! Is this a public value? If not we need to revoke the key and remove the commit.

@alejandrobailo @paabloLC @amitsharm

amitsharm replied:
Key is revoked. @jfagoagas @alejandrobailo @paabloLC

jfagoagas committed:
chore: .env comment

paabloLC commented:
This feature will only be implemented in Prowler Cloud
